Docs/update react tutorial page (#437)
Refinitiv/refinitiv-ui: documents/src/pages/build-app/framework-integration/react.md

# React Guide

@>This guideline uses create-react-app, React v18.2.0

## Using web components in React

> Current version of React does not fully support Web Components. You have to pass the properties and events in a special ways,

Reviewer: ```suggestion
Web components is framework agnostic and it should work with any frameworks in similar fashion as native HTML elements. However, currently, React has not yet fully supported Web components. You are required to create React Component as a wrapper for your Web Component to pass properties and handle custom events.
```

> the best solution is to write a React Component that behaves as a wrapper for your Web Component.

Reviewer: ```suggestion
```

> @>Experimental version of react is fully support Web Components you can try this yourself with a [live demo](https://codesandbox.io/s/tabbar-router-experimental-dq0npp?file=/src/App.js). To follow the status of it check out [custom-elements-everywhere.com](https://custom-elements-everywhere.com/#react).

Reviewer: ```suggestion
@>Experimental version of react is fully support Web Components you can try this yourself with [our live demo](https://codesandbox.io/s/tabbar-router-experimental-dq0npp?file=/src/App.js). To follow the status of it, check out [custom-elements-everywhere.com](https://custom-elements-everywhere.com/#react).
```

> To overcome these shortcomings in our example, we will show how we can create thin React wrapper components around our Web Components. 

Reviewer: ```suggestion
To overcome these shortcomings, in this tutorial, we will show how you can create thin React wrapper components around EF components. 
```

### Create a wrapper component

To create React application we will use [Create React App](https://create-react-app.dev/) CLI tool. Run the following command.

```sh
npx create-react-app my-app
```

> It will create a directory called `my-app` inside the current folder. Inside that directory, it will generate the initial project structure and install the required dependencies.

Reviewer: ```suggestion
At current directory, it will create a new folder called `my-app`. Inside the folder, it will be provided with the initial React project structure and install required dependencies.
```

<br>

> Then installs elements and themes.

Reviewer: ```suggestion
Now, you can install EF components and themes.
```

```sh
npm install @refinitiv-ui/elements @refinitiv-ui/halo-theme
```

Import elements that you want to use and theme in `src/index.js`. You can also import EF components and themes anywhere in react components but for the simplicity we'll import all at once.

```javascript
import '@refinitiv-ui/elements/select';
import '@refinitiv-ui/elements/panel';

import '@refinitiv-ui/halo-theme/dark/imports/native-elements';
import '@refinitiv-ui/elements/select/themes/halo/dark';
import '@refinitiv-ui/elements/panel/themes/halo/dark';
```

<br>

> In our React application, we will need to create a React Select component to wrap our existing `ef-select` component.

Reviewer: ```suggestion
In our React application, we will need to create React `Select` component that wrap `ef-select` component.
```

> We need to map the Web Component properties and events to our React version of the component with the `useLayoutEffect` hook (or `componentDidMount` in class component).

Reviewer: ```suggestion
We need to map `ef-select` properties and events to our `Select` component with the `useLayoutEffect` hook (or `componentDidMount` in class component).
```

```jsx
import React from 'react';

function Select ({ className, value, onChange, data }) {
  const selectRef = React.useRef(); // grab a DOM reference to our `ef-select` 

  React.useLayoutEffect(() => {
    const { current } = ref;

    const handleChange = (event) => {
      onChange(event.detail.value);
    }

    current.data = data;
    current.value = value;
    current.addEventListener('value-changed', handleChange);

    return () => current.removeEventListener('value-changed', handleChange);

  }, [selectRef, onChange, data, value]);

  return <ef-select ref={selectRef} class={className} value={value}></ef-select>
}

export default Select;
```

> One common confusion is that Web Components use `class` instead of `className`.

Reviewer: ```suggestion
One of common confusion when using Web components in React is you need to use `class` instead of `className`.
```

```html
<ef-select class="my-select"></ef-select>
```

<br>

> Add our Select component to `src/App.js`

Reviewer: ```suggestion
Now import `Select` component to `src/App.js`
```

```jsx
import React, { useState } from 'react';
import Select from './Select';
import './App.css';

function App() {
  const [value, setValue] = useState('');

  const data = [{ label: 'Tea', value: 'tea' }, { label: 'Beer', value: 'beer' }];

  const handleChange = (value) => {
    setValue(value);
  }

  const handleClickReset = () => {
    setValue('');
  }

  return (
    <ef-panel spacing>
      <Select className='my-select' onChange={handleChange} data={data} value={value} />
      <button onClick={handleClickReset}>Reset</button>
    </ef-panel>
  );
}

export default App;
```

> And in `src/App.css`

Reviewer: ```suggestion
You can style the component in `src/App.css`.
```

```css
.my-select {
  width: 200px;
  margin-right: 10px;
}
```

Finally, starting your app and it should automatically open `http://localhost:3000/` on your default browser.

```sh
npm start
```

### Using utility wrapper

The previous section has shown you how to wire Web Components into React Components yourself. However, this process could be automated with a wrapper that takes care about formatting objects and arrays to JSON and registering functions as event listeners.

We recommended the package that created by Lit team called [@lit-labs/react](https://github.com/lit/lit/tree/main/packages/labs/react#lit-labsreact).

*> This package is part of [Lit Labs](https://lit.dev/docs/libraries/labs/) that isn't quite ready for production and It's subject to breaking changes.

From inside your project folder, run:

```sh
npm install @lit-labs/react
```

> Import React, a refinitv-ui element class, and createComponent.

Reviewer: ```suggestion
Import React, utility wrapper and EF Select element class. Then you can use `createComponent` to create React wrapper of the element.
```
